### PR TITLE
[Documentation] Add more guides and information on how to customize request and launch configurations

### PR DESCRIPTION
**Pull request by philschmid**
# What does this PR do?

This PR adds more documentation about the launch parameters, how to customize request parameters, stream responses and adds existing guides.

### CODE REVIEW

File: docs/source/basic_tutorials/request_parameters.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Request Parameters for Text Generation Inference
+
+Text Generation Inference support different parameters to control the generation, defining them in the `parameters`` attribute of the payload.
```
Comment (osanseviero) [inline docs/source/basic_tutorials/request_parameters.md:3]:
```suggestion
Text Generation Inference supports different parameters to control the generation, defining them in the `parameters`` attribute of the payload. 
```

File: docs/source/basic_tutorials/request_parameters.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+- `stop`: A list of tokens to stop the generation. The generation will stop when one of the tokens is generated.
+- `top_k`: The number of highest probability vocabulary tokens to keep for top-k-filtering. Default value is `null`, which disables top-k-filtering.
+- `top_p`: The cumulative probability of parameter highest probability vocabulary tokens to keep for nucleus sampling, default to `null`
+- `do_sample`: Whether or not to use sampling; use greedy decoding otherwise. Default value is `false`.
+- `best_of`: Generate best_of sequences and return the one if the highest token logprobs, default to `null`.
+- `details`: Whether or not to return details about the generation. Default value is `false`.
+- `return_full_text`: Whether or not to return the full text or only the generated part. Default value is `false`.
+- `truncate`: Whether or not to truncate the input to the maximum length of the model. Default value is `true`.
+- `typical_p`: The typical probability of a token. Default value is `null`.
+- `watermark`: The watermark to use for the generation. Default value is `false`.
```
Comment (osanseviero) [inline docs/source/basic_tutorials/request_parameters.md:29]: FYI we're working on automating this documentation so it's easy to maintain. The idea is to generate this automatically from the code

File: docs/source/basic_tutorials/stream.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    if r.token.text in gen_kwargs["stop_sequences"]:
+        break
+    # yield the generated token
+    print(r.token.text, end = "")
+    # yield r.token.text
+```
+
+Replace the `print` command with the `yield` or with a function you want to stream the tokens to.
+
+## Streaming requests with JavaScript
```
Comment (osanseviero) [inline docs/source/basic_tutorials/stream.md:52]: Maybe move this to `conceptual/streaming` once it's merged

Comment (osanseviero) [reply docs/source/basic_tutorials/stream.md:52]: Updated to my PR

File: docs/source/launch_parameters.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
+# Launch Parameters for Text Generation Inference
+
+Text Generation Inference allows you to customize the way you serve your models. You can use the following parameters to configure your server. You can enable them by adding them environment variables or by providing them as arguments when running `text-generation-launcher`. Environment variables are in `UPPER_CASE` and arguments are in `lower_case`.
+
+## Model
+
+- `MODEL_ID` - The name of the model to load. Can be a MODEL_ID as listed on huggingface.co like `gpt2` or `OpenAssistant/oasst-sft-1-pythia-12b`. Or it can be a local directory containing the necessary files as saved by `save_pretrained(...)` methods of transformers. Default: `bigscience/bloom-560m`.
```
Comment (Narsil) [inline docs/source/launch_parameters.md:7]: Same here, I'd feel more confortable automating the generation for this.

`text-generation-launcher --help` contains all of those, so we should be able to extract everything from here wdyt ?
To we could have something in the CI that does.


`update-doc --check` (or `update-doc && git diff --is-empty`)

Comment (philschmid) [reply docs/source/launch_parameters.md:7]: Happy to automate but key point we had with customers is they didn't understand whats possible and especially that env vars have a different naming schema. Happy to update if there is an automated way, which is helpful and good.

File: docs/source/basic_tutorials/stream.md (hunk 3 above)
Comment (philschmid) [reply docs/source/basic_tutorials/stream.md]: removed it

File: docs/source/basic_tutorials/request_parameters.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# Request Parameters for Text Generation Inference
+
+Text Generation Inference supports different parameters to control the generation, defining them in the `parameters`` attribute of the payload.
```
Comment (merveenoyan) [inline docs/source/basic_tutorials/request_parameters.md:3]:
```suggestion
Text Generation Inference supports different parameters to control the generation, defining them in the `parameters` attribute of the payload. 
```

File: docs/source/basic_tutorials/request_parameters.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
+# Request Parameters for Text Generation Inference
+
+Text Generation Inference supports different parameters to control the generation, defining them in the `parameters`` attribute of the payload. 
+
+```bash
+curl https://j4xhm53fxl9ussm8.us-east-1.aws.endpoints.huggingface.cloud \
```
Comment (merveenoyan) [inline docs/source/basic_tutorials/request_parameters.md:6]: We always give local serving as URL and we'd like to keep that for consistency I think @osanseviero

Comment (merveenoyan) [inline docs/source/basic_tutorials/request_parameters.md:6]:
```suggestion
curl http://127.0.0.1:8080 \
```

File: docs/source/basic_tutorials/request_parameters.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+-X POST \
+-d '{"inputs":"Once upon a time,", "parameters": {"max_new_tokens": 256}}' \
+-H "Authorization: Bearer <hf_token>" \
+-H "Content-Type: application/json"
+```
+
+
+As of today, the following parameters are supported:
```
Comment (merveenoyan) [inline docs/source/basic_tutorials/request_parameters.md:14]:
```suggestion
As of today, the following parameters are supported:
```
small heads up before moving on with more PRs, for explaining e.g. params we tend to rely on self generating docs (e.g. what you see in swagger)

File: docs/source/basic_tutorials/request_parameters.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+- `temperature`: Controls randomness in the model. Lower values will make the model more deterministic and higher values will make the model more random. Default value is 1.0.
+- `max_new_tokens`: The maximum number of tokens to generate. Default value is 20, max value is 512.
+- `repetition_penalty`: Controls the likelihood of repetition. Default is `null`.
```
Comment (merveenoyan) [inline docs/source/basic_tutorials/request_parameters.md:18]: What other values can it take?

File: docs/source/basic_tutorials/request_parameters.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+- `seed`: The seed to use for random generation. Default is `null`.
+- `stop`: A list of tokens to stop the generation. The generation will stop when one of the tokens is generated.
+- `top_k`: The number of highest probability vocabulary tokens to keep for top-k-filtering. Default value is `null`, which disables top-k-filtering.
+- `top_p`: The cumulative probability of parameter highest probability vocabulary tokens to keep for nucleus sampling, default to `null`
```
Comment (merveenoyan) [inline docs/source/basic_tutorials/request_parameters.md:22]:
```suggestion
- `top_p`: The cumulative probability of parameter highest probability vocabulary tokens to keep for nucleus sampling, defaults to `null`.
```

File: docs/source/guides.md [hunk 10]
```diff
@@ -0,0 +1,6 @@
+# Text Generation Inference Guides
+
+Text Generation Inference (TGI) is integrated into the Hugging Face ecosystem and can be used in many different ways. Below is a list of guides to help you get started with TGI.
```
Comment (merveenoyan) [inline docs/source/guides.md:3]: Maybe this is better to include in either index or README instead of a dedicated section but feel free to dismiss if you disagree

File: docs/source/launch_parameters.md (hunk 4 above)
Comment (merveenoyan) [reply docs/source/launch_parameters.md:7]: @philschmid I'm currently working on it, there are two different notes redirecting folks to use `text-generation-launcher --help` for more in different places across docs. maybe we could resurface that.

Comment (merveenoyan) [inline docs/source/launch_parameters.md:7]:
```suggestion
- `MODEL_ID` - The identifier or path of the model to load. Can be a MODEL_ID as listed on huggingface.co like `gpt2` or `OpenAssistant/oasst-sft-1-pythia-12b`, or it can be a local directory containing the necessary files as saved by `save_pretrained(...)` methods of transformers. Defaults to `bigscience/bloom-560m`.
```

File: docs/source/launch_parameters.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+- `REVISION` - The actual revision of the model if you're referring to a model on Hugging Face Hub. You can use a specific commit id or a branch like `refs/pr/2`.
+
+- `QUANTIZE` - Whether you want the model to be quantized. This will use `bitsandbytes` for quantization on the fly, or `gptq`. 4bit quantization is available through `bitsandbytes` by providing the `bitsandbytes-fp4` or `bitsandbytes-nf4` options. 
+
+- `DTYPE` - The dtype to be forced upon the model. This option cannot be used with `--quantize`.
+
+- `TRUST_REMOTE_CODE` - Whether you want to execute Hub modelling code. Explicitly passing a `revision` is encouraged when loading a model with custom code to ensure no malicious code has been contributed in a newer revision. Default: false
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:15]: Can you make the first sentence clearer? What is Hub modelling code?

File: docs/source/launch_parameters.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+- `DISABLE_CUSTOM_KERNELS` - For some models (like bloom), text-generation-inference implemented custom cuda kernels to speed up inference. Those kernels were only tested on A100. Use this flag to disable them if you're running on different hardware and encounter issues. Default: false
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:17]:
```suggestion
- `DISABLE_CUSTOM_KERNELS` - For some models (like bloom), text-generation-inference includes custom CUDA kernels to speed up inference. Those kernels are only tested on A100. Use this flag to disable them if you encounter issues on different hardware. Default: false
```

File: docs/source/launch_parameters.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+- `ROPE_SCALING` - Rope scaling will only be used for RoPE models and allow rescaling the position rotary to accomodate for larger prompts. Goes together with `rope_factor`. Default: linear
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:19]:
```suggestion
- `ROPE_SCALING` - RoPE scaling will only be used for RoPE models and allow rescaling the position rotary to accommodate for larger prompts. Should be used together with `rope_factor`. Default: linear
```
Also maybe it would be nice for user to refer [here](https://huggingface.co/docs/text-generation-inference/basic_tutorials/preparing_model#rope-scaling) as it's tricky

File: docs/source/basic_tutorials/request_parameters.md [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+- `do_sample`: Whether or not to use sampling; use greedy decoding otherwise. Default value is `false`.
```
Comment (merveenoyan) [inline docs/source/basic_tutorials/request_parameters.md:23]:
```suggestion
- `do_sample`: Whether to use sampling; use greedy decoding otherwise. Default value is `false`.
```

File: docs/source/launch_parameters.md [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+## Inference Settings
+
+- `VALIDATION_WORKERS` - The number of tokenizer workers used for payload validation and truncation inside the router. Default: 2
+
+- `SHARDED` - Whether to shard the model across multiple GPUs. By default text-generation-inference will use all available GPUs to run the model. Setting it to `false` deactivates `num_shard`. 
+
+- `NUM_SHARD` - The number of shards to use if you don't want to use all GPUs on a given machine. You can use `CUDA_VISIBLE_DEVICES=0,1 text-generation-launcher... --num_shard 2` and `CUDA_VISIBLE_DEVICES=2,3 text-generation-launcher... --num_shard 2` to launch 2 copies with 2 shard each on a given machine with 4 GPUs for instance.
+
+- `MAX_CONCURRENT_REQUESTS` - The maximum amount of concurrent requests for this particular deployment. Having a low limit will refuse clients requests instead of having them wait for too long and is usually good to handle backpressure correctly. Default: 128
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:31]:
```suggestion
- `MAX_CONCURRENT_REQUESTS` - The maximum number of concurrent requests for this deployment. Having a low limit will refuse client's requests instead of having them wait for too long and is usually good for handling backpressure correctly. Default: 128
```

File: docs/source/launch_parameters.md [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+- `MAX_BEST_OF` - This is the maximum allowed value for clients to set `best_of`. Best of makes `n` generations at the same time, and return the best in terms of overall log probability over the entire generated sequence. Default: 2
+
+- `MAX_STOP_SEQUENCES` - This is the maximum allowed value for clients to set `stop_sequences`. Stop sequences are used to allow the model to stop on more than just the EOS token, and enable more complex "prompting" where users can preprompt the model in a specific way and define their "own" stop token aligned with their prompt. Default: 4
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:35]:
```suggestion
- `MAX_STOP_SEQUENCES` - This is the maximum allowed value for clients to set `stop_sequences`. Stop sequences are used to allow the model to stop on more than just the EOS token and enable more complex prompting where users can pre-prompt the model in a specific way and define their _own_ stop token aligned with their prompt. Default: 4
```

File: docs/source/launch_parameters.md [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+- `MAX_INPUT_LENGTH` - This is the maximum allowed input length (expressed in number of tokens) for users. The larger this value, the longer prompt users can send which can impact the overall memory required to handle the load. Please note that some models have a finite range of sequence they can handle. Default: 1024
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:37]:
```suggestion
- `MAX_INPUT_LENGTH` - This is users' maximum allowed input length (expressed in the number of tokens). The larger this value, the longer the prompt users can send, which can impact the overall memory required to handle the load. Please note that some models can handle a finite range of sequences. Default: 1024
```

File: docs/source/launch_parameters.md [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+
+- `MAX_TOTAL_TOKENS` - This is the most important value to set as it defines the "memory budget" of running clients requests. Clients will send input sequences and ask to generate `max_new_tokens` on top. with a value of `1512` users can send either a prompt of `1000` and ask for `512` new tokens, or send a prompt of `1` and ask for `1511` max_new_tokens. The larger this value, the larger amount each request will be in your RAM and the less effective batching can be. Default: 2048
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:39]:
```suggestion
- `MAX_TOTAL_TOKENS` - This is the most important value to set as it defines the "memory budget" of running client's requests. Clients will send input sequences and ask to generate `max_new_tokens` on top. With a value of `1512`, users can send either a prompt of `1000` and ask for `512` new tokens, or send a prompt of `1` and ask for `1511` `max_new_tokens`. The larger this value, the larger amount each request will be in your RAM and the less effective batching can be. Default: 2048
```

File: docs/source/launch_parameters.md [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+- `WAITING_SERVED_RATIO` - This represents the ratio of waiting queries vs running queries where you want to start considering pausing the running queries to include the waiting ones into the same batch. `waiting_served_ratio=1.2` Means when 12 queries are waiting and there's only 10 queries left in the current batch we check if we can fit those 12 waiting queries into the batching strategy, and if yes, then batching happens delaying the 10 running queries by a `prefill` run. This setting is only applied if there is room in the batch as defined by `max_batch_total_tokens`. Default: 1.2
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:41]:
```suggestion
- `WAITING_SERVED_RATIO` - This represents the ratio of waiting queries vs running queries where you want to start considering pausing the running queries to include the waiting ones into the same batch. `waiting_served_ratio=1.2` means when 12 queries are waiting, and there are only 10 queries left in the current batch we check if we can fit those 12 waiting queries into the batching strategy, and if yes, then batching happens to delay the ten running queries by a `prefill` run. This setting is only applied if there is room in the batch as defined by `max_batch_total_tokens`. Default: 1.2
```
The sentences are a bit long and hard to follow, can you simplify them please? 🙏

File: docs/source/launch_parameters.md [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+
+- `MAX_BATCH_PREFILL_TOKENS` - Limits the number of tokens for the prefill operation. Since this operation take the most memory and is compute bound, it is interesting to limit the number of requests that can be sent. Default: 4096
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:43]:
```suggestion
- `MAX_BATCH_PREFILL_TOKENS` - Limits the number of tokens for the prefill operation. Since this operation takes the most memory and is compute-bound, it is interesting to limit the number of requests that can be sent. Default: 4096
```
What do you mean when you say: it is interesting to limit the number of requests that can be sent

File: docs/source/launch_parameters.md [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+
+- `MAX_BATCH_TOTAL_TOKENS` - **IMPORTANT** This is one critical control to allow maximum usage of the available hardware. This represents the total amount of potential tokens within a batch. When using padding (not recommended) this would be equivalent of `batch_size` * `max_total_tokens`. However in the non-padded (flash attention) version this can be much finer. For `max_batch_total_tokens=1000`, you could fit `10` queries of `total_tokens=100` or a single query of `1000` tokens. Overall this number should be the largest possible amount that fits the remaining memory (after the model is loaded). Since the actual memory overhead depends on other parameters like if you're using quantization, flash attention or the model implementation, text-generation-inference cannot infer this number automatically.
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:45]:
```suggestion
- `MAX_BATCH_TOTAL_TOKENS` - **IMPORTANT** This is one critical control to allow maximum usage of the available hardware. This represents the total amount of potential tokens within a batch. When using padding (not recommended), this would be equivalent to `batch_size` * `max_total_tokens`. However, this can be much finer in the non-padded (flash attention) version. For `max_batch_total_tokens=1000`, you could fit `10` queries of `total_tokens=100` or a single query of `1000` tokens. Overall this number should be the largest possible amount that fits the remaining memory (after the model is loaded). Since the actual memory overhead depends on other parameters, e.g. if you're using quantization, flash attention or the model implementation, text-generation-inference cannot infer this number automatically.
```
Can you elaborate on: 

> be much finer in the non-padded (flash attention) version

File: docs/source/launch_parameters.md [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+
+- `MAX_WAITING_TOKENS` - This setting defines how many tokens can be passed before forcing the waiting queries to be put on the batch (if the size of the batch allows for it). New queries require 1 `prefill` forward, which is different from `decode` and therefore you need to pause the running batch in order to run `prefill` to create the correct values for the waiting queries to be able to join the batch. With a value too small, queries will always "steal" the compute to run `prefill` and running queries will be delayed by a lot. With a value too big, waiting queries could wait for a very long time before being allowed a slot in the running batch. Default: 20
```
Comment (merveenoyan) [inline docs/source/launch_parameters.md:47]:
```suggestion
- `MAX_WAITING_TOKENS` - This setting defines how many tokens can be passed before forcing the waiting queries to be put on the batch (if the size of the batch allows for it). New queries require 1 `prefill` forward, which is different from `decode`, and therefore, you need to pause the running batch in order to run `prefill` to create the correct values for the waiting queries to be able to join the batch. With a value too small, queries will always "steal" the compute to run `prefill` and running queries will be delayed by a lot. With a value too big, waiting queries could wait for a very long time before being allowed a slot in the running batch. Default: 20
```

**Review by merveenoyan — COMMENTED** [review]
Thanks a lot for this PR!

**Comment by OlivierDehaene** [comment]
What shall we do with this PR? Is it ready for merging?

**Comment by philschmid** [comment]
> What shall we do with this PR? Is it ready for merging?

I stopped looking at it after @osanseviero mentioned

> FYI we're working on automating this documentation so it's easy to maintain. The idea is to generate this automatically from the code
